### PR TITLE
Decompiled ovl_Bg_Haka

### PR DESCRIPTION
**Pull request by Sirius902**

### CODE REVIEW

File: src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c [hunk 1]
```diff
@@ -32,21 +32,146 @@ const ActorInit Bg_Haka_InitVars = {
...
+
+    if (this->dyna.unk_150 != 0.0f) {
+        this->dyna.unk_150 = 0.0f;
+        player->stateFlags2 &= ~0x10;
+    }
+}
+
+void func_8087BAE4(BgHaka* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+    s16 pad;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c:139]:
```suggestion
    s32 pad;
```

File: src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c [hunk 2]
```diff
@@ -32,21 +32,146 @@ const ActorInit Bg_Haka_InitVars = {
...
+void func_8087B7E8(BgHaka* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+
+    if (this->dyna.unk_150 != 0.0f) {
+        if (globalCtx->sceneNum == SCENE_SPOT02 && LINK_IS_CHILD && gSaveContext.nightFlag == 0) {
+            this->dyna.unk_150 = 0.0f;
+            player->stateFlags2 &= ~0x10;
+            if (!Gameplay_InCsMode(globalCtx)) {
+                func_8010B680(globalCtx, 0x5073, NULL);
+                this->dyna.actor.params = 0x64;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c:81]:
```suggestion
                this->dyna.actor.params = 100;
```

File: src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c [hunk 3]
```diff
@@ -32,21 +32,146 @@ const ActorInit Bg_Haka_InitVars = {
...
+                this->actionFunc = func_8087BAE4;
+            }
+        } else if (0.0f < this->dyna.unk_150 ||
+                   (globalCtx->sceneNum == SCENE_SPOT06 && LINK_IS_CHILD && Flags_GetSwitch(globalCtx, 0x23) == 0)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c:85]:
```suggestion
                   (globalCtx->sceneNum == SCENE_SPOT06 && LINK_IS_CHILD && !Flags_GetSwitch(globalCtx, 0x23))) {
```

**Review by Roman971 — APPROVED** [review]

File: src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c [hunk 4]
```diff
@@ -32,21 +32,146 @@ const ActorInit Bg_Haka_InitVars = {
...
+    if (fabsf(sp1C.x) < 34.6f && sp1C.z > -112.8f && sp1C.z < -36.0f) {
+        player->stateFlags2 |= 0x200;
+    }
+}
+
+void func_8087B7E8(BgHaka* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+
+    if (this->dyna.unk_150 != 0.0f) {
+        if (globalCtx->sceneNum == SCENE_SPOT02 && LINK_IS_CHILD && gSaveContext.nightFlag == 0) {
```
Comment (fig02) [inline src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c:76]: !nightFlag

File: src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c [hunk 5]
```diff
@@ -32,21 +32,146 @@ const ActorInit Bg_Haka_InitVars = {
...
+    this->dyna.actor.posRot.pos.x =
+        Math_Sins(this->dyna.actor.posRot.rot.y) * this->dyna.actor.minVelocityY + this->dyna.actor.initPosRot.pos.x;
+    this->dyna.actor.posRot.pos.z =
+        Math_Coss(this->dyna.actor.posRot.rot.y) * this->dyna.actor.minVelocityY + this->dyna.actor.initPosRot.pos.z;
+    if (sp38 != 0) {
+        this->dyna.unk_150 = 0.0f;
+        player->stateFlags2 &= ~0x10;
+        if (this->dyna.actor.params == 1) {
+            func_80078884(NA_SE_SY_CORRECT_CHIME);
+        } else if (gSaveContext.nightFlag != 0 && globalCtx->sceneNum == SCENE_SPOT02) {
```
Comment (fig02) [inline src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c:118]: no !=0 on night flag

File: src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c [hunk 6]
```diff
@@ -32,21 +32,146 @@ const ActorInit Bg_Haka_InitVars = {
...
+    s32 sp38;
+    f32 temp_f0;
+
+    this->dyna.actor.speedXZ += 0.05f;
+    temp_f0 = this->dyna.actor.speedXZ;
+    if (temp_f0 > 1.5f) {
+        this->dyna.actor.speedXZ = 1.5f;
+    } else {
+        this->dyna.actor.speedXZ = temp_f0;
+    }
```
Comment (fig02) [inline src/overlays/actors/ovl_Bg_Haka/z_bg_haka.c:107]: this should be CLAMP_MAX(this->dyna.actor.speedXZ, 1.5f) and should hopefully get rid of the temp.
if the temp is still needed you can name it speedXZ or something

**Pull request merged**